core/services/offchainreporting/data_source.go

```diff
@@ -19,21 +20,38 @@ import (
 // and capturing the result. Additionally, it converts the result to an
 // ocrtypes.Observation (*big.Int), as expected by the offchain reporting library.
 type dataSource struct {
-	pipelineRunner        pipeline.Runner
-	jobSpec               job.Job
-	spec                  pipeline.Spec
-	ocrLogger             logger.Logger
-	runResults            chan<- pipeline.Run
-	currentBridgeMetadata bridges.BridgeMetaData
+	pipelineRunner pipeline.Runner
+	jobSpec        job.Job
+	spec           pipeline.Spec
+	ocrLogger      logger.Logger
+	runResults     chan<- pipeline.Run
+
+	current bridges.BridgeMetaData
+	mu      sync.RWMutex
 }
 
 var _ ocrtypes.DataSource = (*dataSource)(nil)
 
+func (ds *dataSource) updateAnswer(a *big.Int) {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
+	ds.current = bridges.BridgeMetaData{
+		LatestAnswer: a,
+		UpdatedAt:    big.NewInt(time.Now().Unix()),
```

Comment:
This change should resolve the race, but is there still a higher level logical race here in that we are just time-stamping them as they arrive? Is there a different upstream timestamp coupled with this value that we could consider, and then only update when that timestamp is later?

Reply:
those UpdatedAt metadata timestamps represent when the latestAnswer was "computed" (https://github.com/smartcontractkit/external-adapters-js/blob/a35228d5aef735c6ae862c5da8579f6ce07e3a55/README.md#input), which is a bit vague, could include the bridge API responding, the pipeline completing, or just the end of Observe. I think where it is now, end of Observe makes sense as its when we return a value to the OCR protocol, but could see arguments for moving it upstream to run.FinishedAt or even finding the final pipeline task's finished at. In any case, its only the delta between UpdatedAts thats looked at so I don't think it really matters in the current use case

Reply:
This approach seems reasonable to me.